scipy/stats/morestats.py

```diff
@@ -2175,16 +2199,26 @@ def ansari(x, y):
         total = np.sum(a1, axis=0)
         if ind < len(a1)/2.0:
             cind = int(ceil(ind))
-            if ind == cind:
+            if cind != ind:
+                cind = cind - 1
```

Comment:
I thought by "reverting my `cind` change" you meant getting rid of this? 

But don't change anything yet; I am going to take a closer look since there has been so much confusion over this part.

Reply:
Sorry, @tirthasheshpatel I understand what's going on better now. But this code could have used comments when it was initially written, so I think we need to add them in this PR. 

Part of the confusion is that there are two distinct reasons for the `+ 1`s here. It may not be immediately obvious why either one is needed - and much less so when there are two _separate_ reasons and the original and new code achieve them in ways that look different.

This part:
```python3
            cind = int(ceil(ind))
            if cind != ind:
                cind = cind - 1
```
and this part:
```
            find = int(floor(ind))
            if find != ind:
                find = find + 1
```
are just rounding to the nearest `int`. 

The purpose is not obvious, as one would expect `ind` to be an `int` already, since `AB` is an integer and presumably `astart` should be an integer. The purpose is also obscured by the fact that it is done a different way in each branch. Why floor and add in one branch and ceil and subtract in the other?

Now I know there can be difficulties converting `float`s to `int`s:
```
In [16]: np.equal(np.rint(2**54 + 1.0), np.rint(2**54))
Out[16]: True
```
but I have a few questions:

- What precision does `astart` use? `float`s can represent all integers within certain bounds exactly. Is there any reason to suspect that `astart` _won't_ be an exact integer returned as a `float`? If it's so big that not all integers can be represented exactly anymore, we're hosed anyway - careful rounding can't save us.
- Does this three-line version of rounding do anything better than `int(np.round(ind))`, which is much more obvious?
- Should we just use Python arithmetic, which supports arbitrarily large integers, to calculate `astart`, the minimum value of the statistic? (let's not - but we )

(more about the second reason for `+1` later. I have a meeting.)

Reply:
> But this code could have used comments when it was initially written, so I think we need to add them in this PR.

I agree.

> `float`s can represent all integers within certain bounds exactly. Is there any reason to suspect that `astart` _won't_ be an exact integer returned as a `float` - provided that it's within bounds? If it's so big that not all integers can be represented exactly anymore, we're hosed anyway - precision is lost, and careful rounding can't save us.

I also think there is no reason for `astart` not to be an exact integer returned as `float`. If I understand the FORTRAN code correctly, `FPOINT` just takes the integer and converts it to `float` (all the calculations are done as integers so we don't have to worry about getting a `float` during the calculations). So, there is no reason for it to be a `float` which is not an exact integer.

> Does these two three-line rounding routines do something better than `int(np.round(ind))`?

Let me think about this.

> Just a thought: we could just calculate `astart` ourselves using Python arithmetic. It looks like if `m` is the length of `x`, then `astart = floor((m+1)/2)*(1+floor(m/2))`.

I think we will have to use floor division `astart = floor((m+1)//2)*(1+floor(m//2))` because that's what FORTRAN does with integers. If we do that, we already know that `astart` will be an integer and there is no need to `floor` it.

> And if I had written it originally, I would have written a class to represent the distribution (e.g. with methods `cdf` and `sf`). Not sure if we should do it now, but I think it would make the code easier to read.

I agree. This would definitely make it easier to read.

> Let me take a shot at this, actually. We'll see what you think.

Sure!

Reply:
I wasn't thinking clearly before. They round in different directions unless `AB` is already integer. But I still don't understand why they're different from `floor` or ceil`.

Reply:
Now I see that you added those parts I was confused about. Can you explain the intent? How is:
```
            cind = int(ceil(ind))
            if cind != ind:
                cind = cind - 1
```
different from
```python3
cind = int(floor(ind))
```
?

Reply:
OK, I gave this a shot in tirthasheshpatel/scipy#4. I believe the conventions for `cdf` and `sf` in case of ties (non-integer `ind`) is the same as the original code, and I think it makes sense. 

Of course, if we're not going to allow the user to use the `exact` method in case of ties, this is a moot point. `ind` will always be an integer (despite being stored as a float), so it doesn't matter which way we round. But for the record:

When calculating p-values, we want to sum the probability masses from the distribution of the test-statistic under the null hypothesis for all statistic values _equally extreme or more extreme_ than the observed value of the statistic. _Because the exact distribution implemented here does not consider ties, the probability mass at non-integer values of the statistic is 0._ Therefore, when the observed value of the statistic is non-integer, we sum over statistic values _strictly more extreme than_ the observed value. When summing over the left tail, this means we sum from the left up to `floor(ind)`. When summing over the right tail, this means we sum from `ceil(ind)` to the right. That is effectively what the original code does, so I've preserved that behavior.

Reply:
If you prefer, we can also leave this as-is, now that I understand what's going on. But if you like the structure of the PR to your branch and agree that it would help with the exact test, feel free to use it.

Reply:
> OK, I gave this a shot in [tirthasheshpatel#4](https://github.com/tirthasheshpatel/scipy/pull/4). I believe the conventions for `cdf` and `sf` in case of ties (non-integer `ind`) is the same as the original code, and I think it makes sense.

I looked at it and it's very helpful, thanks!

> When calculating p-values, we want to sum the probability masses from the distribution of the test-statistic under the null hypothesis for all statistic values _equally extreme or more extreme_ than the observed value of the statistic. _Because the exact distribution implemented here does not consider ties, the probability mass at non-integer values of the statistic is 0._ Therefore, when the observed value of the statistic is non-integer, we sum over statistic values _strictly more extreme than_ the observed value. When summing over the left tail, this means we sum from the left up to and including `floor(ind)`. When summing over the right tail, this means we sum from `ceil(ind)` to the right. That is effectively what the original code does, so I've preserved that behavior.

All of this makes sense to me.

> If you prefer, we can also leave this as-is, now that I understand what's going on. But if you like the structure of the PR to your branch and agree that it would help with the exact test, feel free to merge it.

I think that tirthasheshpatel#4 is way more readable and the comments help a lot. So, I think it would be very nice to get it merged in this branch!